File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_games/lib/src/state/games_state.dart
import '../models/games_models.dart';

enum GamesStatus { idle, loading, ready, error }

class GamesState {
  const GamesState({
    this.status = GamesStatus.idle,
    this.lastResult,
    this.error,
  });

  final GamesStatus status;
  final GamesPageResult<Object?>? lastResult;
  final String? error;

  GamesState loading() => const GamesState(status: GamesStatus.loading);

  GamesState loaded(GamesPageResult<Object?> result) =>
      GamesState(status: GamesStatus.ready, lastResult: result);

  GamesState failed(String message) =>
      GamesState(status: GamesStatus.error, error: message);
}

File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-games/src/state/gamesState.ts
import type { GamesPageResult } from "../types/gamesModels.js";

export interface GamesState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: GamesPageResult<unknown>;
  readonly error?: string;
}

export function initialGamesState(): GamesState {
  return { status: "idle" };
}

export function reduceGamesState(
  state: GamesState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: GamesPageResult<unknown> }
    | { type: "failed"; error: string },
): GamesState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}
